region/client_test.go

```diff
@@ -1280,6 +1285,202 @@ func TestMarshalJSONNilValues(t *testing.T) {
 	}
 }
 
+func TestScanConcurrencyControl(t *testing.T) {
+	// Create a client with scan concurrency control enabled
+	done := make(chan struct{})
+	defer close(done)
+
+	// Create token bucket with capacity of 1 and initially 1 token
+	tokenBucket, err := NewToken(1, 1, done)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c := &client{
+		sent:            make(map[uint32]hrpc.Call),
+		scanTokenBucket: tokenBucket,
+		logger:          slog.Default(),
+	}
+
+	ctx := context.Background()
+
+	// Create first scan request
+	scan1, err := hrpc.NewScanRange(ctx, []byte("table"), []byte("start"), []byte("stop"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create second scan request
+	scan2, err := hrpc.NewScanRange(ctx, []byte("table"), []byte("start"), []byte("stop"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Register first scan - should succeed immediately since we have 1 token
+	id1, err := c.registerRPC(scan1)
+	if err != nil {
+		t.Fatalf("First scan registration should succeed, got error: %v", err)
+	}
+
+	// Channel to track when the second registration starts and completes
+	registrationStarted := make(chan struct{})
+	registrationCompleted := make(chan struct{})
+
+	var id2 uint32
+	var registerErr error
+
+	// Start goroutine to register second scan
+	go func() {
+		close(registrationStarted)
+		id2, registerErr = c.registerRPC(scan2)
+		close(registrationCompleted)
+	}()
+
+	// Wait for the registration to start
+	<-registrationStarted
+
+	// Give the registration some time to potentially complete if it wasn't blocking
+	select {
+	case <-registrationCompleted:
+		t.Fatal("Second scan registration should be blocked")
+	case <-time.After(100 * time.Millisecond):
+		// Good - registration is blocked as expected
+	}
+
+	// Now unregister the first scan to free up a token
+	rpc := c.unregisterRPC(id1)
+	if rpc != scan1 {
+		t.Fatalf("Expected to get back scan1, got %v", rpc)
+	}
+
+	// The second registration should now complete
+	select {
+	case <-registrationCompleted:
+		// Good - registration unblocked
+		if registerErr != nil {
+			t.Fatalf("Second scan registration failed: %v", registerErr)
+		}
+		if id2 == 0 {
+			t.Fatal("Second scan registration should have assigned a non-zero ID")
+		}
+	case <-time.After(1 * time.Second):
+		t.Fatal("Second scan registration should have unblocked after first was unregistered")
```

Comment:
This timeout can cause flakiness in CI runners, I suggest removing it. `go test` has a built in timeout, so no need to define your own.

And if you use synctest then we can deterministically test this without any sleeps required.